server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1668,56 +1806,214 @@ private static boolean isIndexPresent(ClusterService clusterService, String inde
-                        final Map<String, BlobContainer> shardBlobs = indexEntry.getValue().children();
-                        for (Map.Entry<String, BlobContainer> shardBlob : shardBlobs.entrySet()) {
-                            for (String blob : shardBlob.getValue().listBlobs().keySet()) {
-                                final Optional<String> snapshotUUID = extractShallowSnapshotUUID(blob);
-                                if (snapshotUUID.isPresent()) {
-                                    releaseRemoteStoreLockAndCleanup(
-                                        shardBlob.getKey(),
-                                        snapshotUUID.get(),
-                                        shardBlob.getValue(),
-                                        remoteStoreLockManagerFactory
-                                    );
-                                }
-                            }
-                        }
-                    }
-                    // Deleting the index folder
-                    deleteResult = indexEntry.getValue().delete();
-                    logger.debug("[{}] Cleaned up stale index [{}]", metadata.name(), indexSnId);
-                } catch (IOException e) {
-                    logger.warn(
-                        () -> new ParameterizedMessage(
-                            "[{}] index {} is no longer part of any snapshots in the repository, "
-                                + "but failed to clean up their index folders",
-                            metadata.name(),
-                            indexSnId
-                        ),
-                        e
-                    );
-                } catch (Exception e) {
-                    assert false : e;
-                    logger.warn(new ParameterizedMessage("[{}] Exception during single stale index delete", metadata.name()), e);
+        if (indexEntry == null) {
+            return;
+        }
+        final String indexSnId = indexEntry.getKey();
+        threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(ActionRunnable.supply(listener, () -> {
+            try {
+                logger.debug("[{}] Found stale index [{}]. Cleaning it up", metadata.name(), indexSnId);
+                List<String> matchingShardPaths = findMatchingShardPaths(indexSnId, snapshotShardPaths);
+                Optional<String> highestGenShardPaths = findHighestGenerationShardPaths(matchingShardPaths);
+                ShardInfo shardInfo = getShardInfo(highestGenShardPaths, idToShardInfoMap, indexSnId);
+
+                if (remoteStoreLockManagerFactory != null) {
+                    cleanupRemoteStoreLocks(indexEntry, shardInfo, remoteStoreLockManagerFactory);
                 }
 
-                executeOneStaleIndexDelete(staleIndicesToDelete, remoteStoreLockManagerFactory, listener);
+                // The below deleteShardData would be no-op for FIXED path type snapshots. Otherwise, this deletes the
```

Comment:
why ? Won't it delete all the shard level paths for fixed path snapshots as well. ?

Reply:
My bad - This was true in 3-4 commits older revision. Let me update it now. As of now, it is capable of cleaning up shard data for all path types.